### DISCUSSION TITLE
what are the default serialization steps applied by fastapi?

### DISCUSSION THREAD

**Discussion by orfisko** [fastapi/fastapi#12232] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# not applicable
```

### Description

Currently I am facing a problem when I want to send a large response. It takes quite long to have it processed by fastapi and I am investigating why. 

The problem I currently face to get this done quickly is that it is impossible to find out what the default serialization steps are that fastapi is applying when using the response_model thing in the router. A lot of old information is on the internet, with the introduction of pydantic v2 a lot has changed.  
The questions I have: 
1. In the documentation there is still reference to jsonable_encoder being used for serialization, but I doubt if that is still being used these days by default?
2. If my response model is of the same type as the object I am returning (i.e. also a pydantic model), is there still validation being applied by fastapi? If so, can I disable that? 
3. Can I circumvent the recreation of the model by using a direct response and put a foo.model_dump(mode="json") in the payload? How can I still keep the documentation in that case?

I did a quick test where I removed the response model:

`
from fastapi.responses import JSONResponse
@router.get(
    "",
    dependencies=[Depends(Authorizer())],
)
`
and returned my payload like this:
`
return JSONResponse(content=[eboard.model_dump(mode="json") for eboard in eboards])
`
but that barely makes a difference in terms of performance... So I am wandering what I am actually skipping by implementing it as such.

### Operating System

Linux

### FastAPI Version

0.115

### Pydantic Version

2.9.2

### Python Version

3.11.10

**Comment by YuriiMotov** [comment, marked as answer]
> In the documentation there is still reference to jsonable_encoder being used for serialization, but I doubt if that is still being used these days by default?

Yes, `jsonable_encoder` is still used

> If my response model is of the same type as the object I am returning (i.e. also a pydantic model), is there still validation being applied by fastapi? If so, can I disable that?

Yes, validation will be applied even if you return an instance of the same model as set as response model.
The only way to bypass this validation is to return an instance of `Response` class (or sub-classes) directly from your endpoint.

> Can I circumvent the recreation of the model by using a direct response and put a foo.model_dump(mode="json") in the payload? How can I still keep the documentation in that case?

Specify the `response_model` parameter:

```python
@app.put("/items/{id}", response_model=Item)
def update_item(id: str, item: Item):
    json_compatible_item_data = jsonable_encoder(item)
    return JSONResponse(content=json_compatible_item_data)
```
